### PR TITLE
Generate username automatically - new signup

### PR DESCRIPTION
**Pull request by lakshayman**
Date: 18-08-2024

Developer Name: @lakshayman 

---

## Description

We want to generate the username automatically, so we removed the step and we are generating the username while submitting.

### Documentation Updated?

- [ ] Yes
- [x] No

### Under Feature Flag

- [ ] Yes
- [x] No

### Database Changes

- [ ] Yes
- [x] No

### Breaking Changes

- [x] Yes
- [ ] No

### Development Tested?

- [ ] Yes
- [x] No

## Screenshots
<details>
<summary>Screenshot 1</summary>

https://github.com/user-attachments/assets/0d278e80-8498-4b48-8fc9-c67bd8d96d72

</details>

## Test Coverage
<details>
<summary>Screenshot 1</summary>

image

</details>

### CODE REVIEW

File: app/controllers/new-signup.js [hunk 1]
```diff
@@ -91,10 +122,14 @@ export default class NewSignUpController extends Controller {
...
   @action async signup() {
+    const user = await this.generateUsername(
+      this.signupDetails.firstName,
+      this.signupDetails.lastName
+    );
     const signupDetails = {
       first_name: this.signupDetails.firstName,
       last_name: this.signupDetails.lastName,
-      username: this.signupDetails.username,
+      username: user?.username,
```
Comment (prakashchoudhary07) [inline app/controllers/new-signup.js:132]: User cannot be optional

Comment (prakashchoudhary07) [reply app/controllers/new-signup.js:132]: how and when will the users be optional

Comment (lakshayman) [reply app/controllers/new-signup.js:132]: In case of the API error, it is possible that user is undefined.

File: app/controllers/new-signup.js [hunk 2]
```diff
@@ -41,6 +43,35 @@ export default class NewSignUpController extends Controller {
     this.analytics.trackEvent(NEW_SIGNUP_FLOW.USER_GETTING_STARTED);
   }
 
+  async generateUsername(firstname, lastname) {
```
Comment (iamitprakash) [inline app/controllers/new-signup.js:46]: minor fix for return statements

```

async generateUsername(firstname, lastname) {
    if (typeof firstname !== 'string' || typeof lastname !== 'string') {
        throw new Error('Invalid input: firstname and lastname must be strings');
    }

    try {
        const sanitizedFirstname = firstname.toLowerCase();
        const sanitizedLastname = lastname.toLowerCase();
        const response = await fetch(
            `${ENV.BASE_API_URL}/users/username?dev=true&firstname=${sanitizedFirstname}&lastname=${sanitizedLastname}`,
            {
                method: 'GET',
                headers: {
                    'Content-Type': 'application/json',
                },
                credentials: 'include',
            }
        );
        const user = await response.json();
        if (user && user.username) {
            return user;
        }
        throw new Error('Username generation failed: Invalid response from server');
    } catch (err) {
        this.toast.error(
            ERROR_MESSAGES.usernameGeneration,
            'error!',
            toastNotificationTimeoutOptions
        );
        throw err; // Re-throw the error after showing the toast
    }
}
````

**Review by iamitprakash — CHANGES_REQUESTED** [review]

File: app/controllers/new-signup.js (hunk 1 above)
Comment (prakashchoudhary07) [reply app/controllers/new-signup.js:132]: If the users is undefined should we come till this flow?

Comment (iamitprakash) [reply app/controllers/new-signup.js:132]: > If the users is undefined should we come till this flow?

for me no @lakshayman

File: app/controllers/new-signup.js (hunk 2 above)
Comment (lakshayman) [reply app/controllers/new-signup.js:47]: Resolved this in the new commit please check

File: app/controllers/new-signup.js (hunk 1 above)
Comment (lakshayman) [reply app/controllers/new-signup.js]: Updated the code, according to it. Please check @iamitprakash @prakashchoudhary07

**Review by prakashchoudhary07 — APPROVED** [review]

**Review by iamitprakash — APPROVED** [review]

**Pull request merged**